utils_cv/detection/model.py

```diff
@@ -0,0 +1,283 @@
+        """ create training and validation data loaders
+        """
+        self.train_ds, self.test_ds = dataset.split_train_test(
+            train_ratio=train_ratio
+        )
+
+        self.train_dl = DataLoader(
+            self.train_ds,
+            batch_size=self.batch_size,
+            shuffle=True,
+            num_workers=4,
+            collate_fn=collate_fn,
+        )
+
+        self.test_dl = DataLoader(
+            self.test_ds,
+            batch_size=self.batch_size,
+            shuffle=False,
+            num_workers=4,
+            collate_fn=collate_fn,
+        )
+
+    def fit(self, epochs: int, print_freq: int = 10) -> None:
+        """ The main training loop. """
+        self.losses = []
+        self.epochs = epochs
+        for epoch in range(self.epochs):
+            # train for one epoch, printing every 10 iterations
+            logger = train_one_epoch(
+                self.model,
+                self.optimizer,
+                self.train_dl,
+                self.device,
+                epoch,
+                print_freq=print_freq,
+            )
+            self.losses.append(logger.meters["loss"].median)
+
+            # update the learning rate
+            self.lr_scheduler.step()
+
+            # evaluate
+            self.evaluate(dl=self.train_dl)
+
+    def plot_losses(self, figsize: Tuple[int, int] = (10, 5)) -> None:
+        """ Plot training loss from fitting. """
+        fig = plt.figure(figsize=figsize)
+        ax = fig.add_subplot(111)
+        ax.set_xlim([0, self.epochs - 1])
+        ax.set_xticks(range(0, self.epochs))
+        ax.set_title("Loss over epochs")
+        ax.set_xlabel("epochs")
+        ax.set_ylabel("loss")
+        ax.plot(self.losses)
+
+    def evaluate(self, dl: DataLoader = None) -> CocoEvaluator:
+        """ eval code on validation/test set. """
+        if dl is None:
+            dl = self.test_dl
+        return evaluate(self.model, dl, device=self.device)
+
+    def get_model(self) -> nn.Module:
+        """ returns the model object. """
+        return self.model
+
+    def inference(self, im_path: Union[str, Path]):
+        """ Performs inferencing on an image path. """
+        im = Image.open(im_path)
+        transform = transforms.Compose([transforms.ToTensor()])
+        im = transform(im).cuda()
+        model = self.get_model().eval()  # eval mode
+        with torch.no_grad():
+            pred = model([im])
+        return pred
+
+    def show_detection_vs_ground_truth(self, idx: int, ax: plt.axes) -> None:
+        """ Plots the bounding box predictions and ground truth
+
+        Args:
+            idx: the index of the dataset to visualize
+            ax: axes to draw on
+
+        Returns nothing but plots graph.
+        """
+        im_path = (
+            self.dataset.root
+            / self.dataset.image_folder
+            / self.dataset.ims[idx]
+        )
+        im = cv2.imread(str(im_path))
```

Comment:
Looks like we can get rid of cv2 by loading using PIL here instead? Out of interest, the ax.imshow() does that show the image in flipped RGB order?